File: ruby/lib/dubs/generator.rb
# frozen_string_literal: true

module Dubs
  class Generator
    DEFAULTS = {
      theme: nil,
      category: nil,
      pattern: :default,
      token: :numeric,
      token_length: 4,
      token_case: :lower,
      separator: "-",
      adjective_group: :general,
      seed: nil
    }.freeze

    def initialize(**options)
      @options = DEFAULTS.merge(options)
    end

    def generate
      theme = resolve_theme
      adjective = pick_adjective
      noun = pick_noun(theme)
      token_value = generate_token

      template = Pattern.resolve(@options[:pattern])
      Pattern.interpolate(template, adjective: adjective, noun: noun, token: token_value, separator: @options[:separator])
    end

    private

    def resolve_theme
      name = @options[:theme]
      name ? Data.theme(name) : Data.random_theme
    end

    def pick_adjective
      adjectives = Data.adjectives(@options[:adjective_group])
      if @options[:seed]
        adjectives[@options[:seed] % adjectives.length]
      else
        adjectives.sample
      end
    end

    def pick_noun(theme)
      nouns = theme.nouns(@options[:category])
      if @options[:seed]
        adj_count = Data.adjectives(@options[:adjective_group]).length
        nouns[(@options[:seed] / adj_count) % nouns.length]
      else
        nouns.sample
      end
    end

    def generate_token
      Token.generate(
        type: @options[:token],
        length: @options[:token_length],
        token_case: @options[:token_case],
        seed: @options[:seed]
      )
    end
  end
end

File: rust/src/generator.rs
use crate::data::ADJECTIVES;
use crate::pattern::{self, PatternName};
use crate::theme::Theme;
use crate::token::{self, TokenCase, TokenType};

/// Builder for generating themed names.
///
/// ```
/// use dubs::{Dubs, Theme, TokenType, TokenCase, PatternName};
///
/// let name = Dubs::new(Theme::by_name("gundam").unwrap())
///     .token_type(TokenType::Hex)
///     .token_length(6)
///     .token_case(TokenCase::Upper)
///     .pattern(PatternName::Designation)
///     .generate();
///
/// assert!(name.contains("-class-"));
/// ```
pub struct Dubs {
    theme: Theme,
    category: Option<String>,
    pattern: PatternName,
    token_type: TokenType,
    token_length: usize,
    token_case: TokenCase,
    separator: String,
    adjective_group: String,
    seed: Option<u64>,
}

impl Dubs {
    /// Create a new generator with a theme.
    pub fn new(theme: Theme) -> Self {
        Self {
            theme,
            category: None,
            pattern: PatternName::Default,
            token_type: TokenType::Numeric,
            token_length: 4,
            token_case: TokenCase::Lower,
            separator: "-".to_string(),
            adjective_group: "general".to_string(),
            seed: None,
        }
    }

    pub fn category(mut self, category: &str) -> Self {
        self.category = Some(category.to_string());
        self
    }

    pub fn pattern(mut self, pattern: PatternName) -> Self {
        self.pattern = pattern;
        self
    }

    pub fn token_type(mut self, token_type: TokenType) -> Self {
        self.token_type = token_type;
        self
    }

    pub fn token_length(mut self, length: usize) -> Self {
        self.token_length = length;
        self
    }

    pub fn token_case(mut self, case: TokenCase) -> Self {
        self.token_case = case;
        self
    }

    pub fn separator(mut self, sep: &str) -> Self {
        self.separator = sep.to_string();
        self
    }

    pub fn adjective_group(mut self, group: &str) -> Self {
        self.adjective_group = group.to_string();
        self
    }

    pub fn seed(mut self, seed: u64) -> Self {
        self.seed = Some(seed);
        self
    }

    /// Generate the name.
    pub fn generate(&self) -> String {
        let adjective = self.pick_adjective();
        let noun = self.pick_noun();
        let tok = token::generate(self.token_type, self.token_length, self.token_case, self.seed);
        let template = self.pattern.template();

        pattern::interpolate(template, &adjective, &noun, &tok, &self.separator)
    }

    fn pick_adjective(&self) -> String {
        let adjectives = ADJECTIVES
            .group(&self.adjective_group)
            .unwrap_or(&ADJECTIVES.general);

        match self.seed {
            Some(s) => adjectives[(s as usize) % adjectives.len()].clone(),
            None => {
                use rand::prelude::IndexedRandom;
                adjectives
                    .choose(&mut rand::rng())
                    .cloned()
                    .unwrap_or_default()
            }
        }
    }

    fn pick_noun(&self) -> String {
        let cat = self.category.as_deref();
        let nouns = self.theme.nouns(cat).unwrap_or_default();

        match self.seed {
            Some(s) => {
                let adj_len = ADJECTIVES
                    .group(&self.adjective_group)
                    .map(|a| a.len())
                    .unwrap_or(1);
                nouns[((s as usize) / adj_len) % nouns.len()].clone()
            }
            None => {
                use rand::prelude::IndexedRandom;
                nouns
                    .choose(&mut rand::rng())
                    .cloned()
                    .unwrap_or_default()
            }
        }
    }
}
